### PR TITLE
Fix proposal for backend start up (issue #23)

### LINKED ISSUE #23: Node register app can fail
Sometimes, when we start substra-backend with docker-compose (cf `docker/start.py`) with `dev` settings, the node-register app can fail (see logs after) and block the container as `runserver` will not exit.

In production, we start substra-backend with `uwsgi`  which has a `need-app` parameters that prevent from this issue

```
INFO - 2019-11-04 13:43:30,885 - events.apps - Start the event application.
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/hfc/fabric/client.py", line 1711, in chaincode_invoke
    timeout=wait_for_event_timeout)
  File "/usr/local/lib/python3.6/asyncio/tasks.py", line 362, in wait_for
    raise futures.TimeoutError()
concurrent.futures._base.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/app/substrapp/ledger_utils.py", line 159, in call_ledger
    response = loop.run_until_complete(chaincode_calls[call_type](**params))
  File "/usr/local/lib/python3.6/asyncio/base_events.py", line 484, in run_until_complete
    return future.result()
  File "/usr/local/lib/python3.6/site-packages/hfc/fabric/client.py", line 1721, in chaincode_invoke
    raise TimeoutError('waitForEvent timed out.')
TimeoutError: waitForEvent timed out.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "manage.py", line 15, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 357, in execute
    django.setup()
  File "/usr/local/lib/python3.6/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python3.6/site-packages/django/apps/registry.py", line 120, in populate
    app_config.ready()
  File "/usr/src/app/node-register/apps.py", line 10, in ready
    invoke_ledger(fcn='registerNode', args=[''], sync=True)
exception calling callback for <Future at 0x7f60aeeb2cc0 state=finished raised _Rendezvous>
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.6/site-packages/aiogrpc/utils.py", line 126, in _next
    return next(self._iterator)
  File "/usr/local/lib/python3.6/site-packages/grpc/_channel.py", line 388, in __next__
    return self._next()
  File "/usr/local/lib/python3.6/site-packages/grpc/_channel.py", line 382, in _next
    raise self
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
        status = StatusCode.CANCELLED
        details = "Locally cancelled by application!"
        debug_error_string = "None"
>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/concurrent/futures/_base.py", line 324, in _invoke_callbacks
    callback(self)
  File "/usr/local/lib/python3.6/asyncio/futures.py", line 417, in _call_set_state
    dest_loop.call_soon_threadsafe(_set_state, destination, source)
  File "/usr/local/lib/python3.6/asyncio/base_events.py", line 637, in call_soon_threadsafe
    self._check_closed()
  File "/usr/local/lib/python3.6/asyncio/base_events.py", line 377, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed  File "/usr/src/app/substrapp/ledger_utils.py", line 90, in _wrapper

    return fn(*args, **kwargs)
  File "/usr/src/app/substrapp/ledger_utils.py", line 208, in invoke_ledger
    response = call_ledger('invoke', fcn=fcn, args=args, kwargs=params)
  File "/usr/src/app/substrapp/ledger_utils.py", line 161, in call_ledger
    raise LedgerTimeout(str(e))
substrapp.ledger_utils.LedgerTimeout: waitForEvent timed out.
exception calling callback for <Future at 0x7f60aef01cf8 state=finished raised _Rendezvous>
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.6/site-packages/aiogrpc/utils.py", line 126, in _next
    return next(self._iterator)
  File "/usr/local/lib/python3.6/site-packages/grpc/_channel.py", line 388, in __next__
    return self._next()
  File "/usr/local/lib/python3.6/site-packages/grpc/_channel.py", line 382, in _next
    raise self
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
        status = StatusCode.CANCELLED
        details = "Locally cancelled by application!"
        debug_error_string = "None"
>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/concurrent/futures/_base.py", line 324, in _invoke_callbacks
    callback(self)
  File "/usr/local/lib/python3.6/asyncio/futures.py", line 417, in _call_set_state
    dest_loop.call_soon_threadsafe(_set_state, destination, source)
  File "/usr/local/lib/python3.6/asyncio/base_events.py", line 637, in call_soon_threadsafe
    self._check_closed()
  File "/usr/local/lib/python3.6/asyncio/base_events.py", line 377, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed

```

### PR DESCRIPTION
**Pull request by Kelvin-M**
Fix #23

### CODE REVIEW

File: backend/node-register/apps.py [hunk 1]
```diff
@@ -1,10 +1,31 @@
...
-        # args is set to empty because fabric-sdk-py doesn't allow None args for invoke operations
-        invoke_ledger(fcn='registerNode', args=[''], sync=True)
+
+        # We try until success if dev mode, if it fails the backend will not start
+        # In production we let this app fails
+        while True:
+            try:
+                # args is set to empty because fabric-sdk-py doesn't allow None args for invoke operations
+                invoke_ledger(fcn='registerNode', args=[''], sync=True)
+            except Exception as e:
```
Comment (samlesu) [inline backend/node-register/apps.py]: Could we catch a specific exception? Like a `TimeoutError` or a `GrpcError` for instance (the traceback is quite hard to follow).

Comment (Kelvin-M) [reply backend/node-register/apps.py]: Yes, I will put `RpcError` in the except catch.

File: backend/node-register/apps.py [hunk 2]
```diff
@@ -1,10 +1,31 @@
...
+                logger.exception(e)
+
+                if not settings.DEBUG:
+                    break
```
[the remark this answers is not in the document]
Comment (Kelvin-M) [reply backend/node-register/apps.py]: Yes, It will be cleaner !

File: backend/node-register/apps.py [hunk 3]
```diff
@@ -1,10 +1,31 @@
...
+
+                time.sleep(5)
+                logger.info('Retry to register the node to the ledger')
+            else:
+                break
```
[the remark this answers is not in the document]
Comment (Kelvin-M) [reply backend/node-register/apps.py]: I would like to add a log to show that something happen.
So I will need to keep this break value :)

[the remark this answers is not in the document]
Comment (samlesu) [reply backend/node-register/apps.py]: 👍

File: backend/node-register/apps.py [hunk 4]
```diff
@@ -1,10 +1,25 @@
...
+        while True:
+            try:
+                # args is set to empty because fabric-sdk-py doesn't allow None args for invoke operations
+                invoke_ledger(fcn='registerNode', args=[''], sync=True)
+            except Exception as e:
+                logger.exception(e)
+                time.sleep(5)
+                logger.info('Retry to register the node to the ledger')
+            else:
+                break
```
[the remark this answers is not in the document]
Comment (Kelvin-M) [reply backend/node-register/apps.py]: Yes, I raise in production :)

**Comment by Kelvin-M** [comment]
@samlesu Is the current proposal is good for you ? I tried to combine your review with improvements :)
